firedrake/formmanipulation.py

```diff
@@ -147,6 +148,29 @@ def cofunction(self, o):
         else:
             return Cofunction(W, val=MixedDat(o.dat[i] for i in indices))
 
+    def matrix(self, o):
+        ises = []
+        args = []
+        for a in o.arguments():
+            V = a.function_space()
+            iset = PETSc.IS()
+            if a.number() in self.blocks:
+                asplit = self._subspace_argument(a)
+                for f in self.blocks[a.number()]:
+                    fset = V.dof_dset.field_ises[f]
+                    iset = iset.expand(fset)
+            else:
+                asplit = a
+                for fset in V.dof_dset.field_ises:
+                    iset = iset.expand(fset)
+
+            ises.append(iset)
+            args.append(asplit)
+
+        submat = o.petscmat.createSubMatrix(*ises)
+        bcs = ()
+        return AssembledMatrix(tuple(args), bcs, submat)
```

Comment:
Why enforce this to be `AssembledMatrix`? Why not `type(o)` so that the matrix type is maintained?

Reply:
`ImplicitMatrixContext.createSubMatrix` already does that for you.